File: test_spark_course_sync.py
"""
Test SPARK Course Sync Webhook

Tests the /spark/webhook/course endpoint to verify:
1. Quest creation from course data
2. Sample task creation from assignments
3. HMAC signature validation

Usage:
    python test_spark_course_sync.py [--prod]

Environment variables required:
    SPARK_WEBHOOK_SECRET - HMAC secret for signing webhook payloads
"""

import os
import sys
import json
import hmac
import hashlib
import requests
from datetime import datetime

# Configuration
if '--prod' in sys.argv:
    BACKEND_URL = 'https://optio-prod-backend.onrender.com'
    print("Using PRODUCTION backend")
else:
    BACKEND_URL = 'https://optio-dev-backend.onrender.com'
    print("Using DEVELOPMENT backend")

SPARK_WEBHOOK_SECRET = os.getenv('SPARK_WEBHOOK_SECRET')

if not SPARK_WEBHOOK_SECRET:
    print('Error: SPARK_WEBHOOK_SECRET environment variable not set')
    print('\nTo set it:')
    print('  export SPARK_WEBHOOK_SECRET=<secret>')
    print('\nGet the secret from:')
    print('  Render Dashboard → optio-dev-backend → Environment → SPARK_WEBHOOK_SECRET')
    sys.exit(1)

# Test course data with assignments
course_data = {
    "spark_org_id": "24",  # Demo org ID from SPARK developer
    "spark_course_id": "test_course_biology_101",
    "course_title": "Biology 101 - Test Course",
    "course_description": "Introduction to biology - automated test course with assignments",
    "assignments": [
        {
            "spark_assignment_id": "assign_bio_001",
            "title": "Cell Structure Lab Report",
            "description": "Complete the cell structure lab and submit a detailed report with diagrams",
            "due_date": "2025-02-15T23:59:59Z"
        },
        {
            "spark_assignment_id": "assign_bio_002",
            "title": "Photosynthesis Essay",
            "description": "Write a 500-word essay explaining the process of photosynthesis",
            "due_date": "2025-02-22T23:59:59Z"
        },
        {
            "spark_assignment_id": "assign_bio_003",
            "title": "Ecosystem Project",
            "description": "Create a presentation about a local ecosystem",
            "due_date": "2025-03-01T23:59:59Z"
        }
    ]
}


def calculate_signature(payload):
    """Calculate HMAC-SHA256 signature for webhook payload"""
    # Sort keys for consistent signature
    payload_str = json.dumps(payload, separators=(',', ':'), sort_keys=True)
    signature = hmac.new(
        SPARK_WEBHOOK_SECRET.encode(),
        payload_str.encode(),
        hashlib.sha256
    ).hexdigest()
    return signature


def test_course_sync_webhook():
    """Test the course sync webhook endpoint"""
    print('\nTesting SPARK Course Sync Webhook')
    print('=' * 50)
    print(f'\nBackend URL: {BACKEND_URL}')
    print(f'Course ID: {course_data["spark_course_id"]}')
    print(f'Course Title: {course_data["course_title"]}')
    print(f'Assignment Count: {len(course_data["assignments"])}')
    print('\nAssignments:')
    for idx, assignment in enumerate(course_data['assignments'], 1):
        print(f'  {idx}. {assignment["title"]} (ID: {assignment["spark_assignment_id"]})')

    # Calculate signature
    signature = calculate_signature(course_data)
    print(f'\nCalculated HMAC signature: {signature[:20]}...')

    # Make request
    print(f'\nSending POST request to /spark/webhook/course...\n')

    try:
        response = requests.post(
            f'{BACKEND_URL}/spark/webhook/course',
            headers={
                'Content-Type': 'application/json',
                'X-Spark-Signature': signature
            },
            json=course_data,
            timeout=30
        )

        print(f'Response Status: {response.status_code}')

        try:
            response_data = response.json()
            print(f'Response Body: {json.dumps(response_data, indent=2)}')
        except json.JSONDecodeError:
            print(f'Response Text: {response.text}')
            response_data = {}

        if response.ok:
            print('\n✓ SUCCESS - Course sync webhook processed successfully')
            if 'quest_id' in response_data:
                print(f'  Quest ID: {response_data["quest_id"]}')
            if 'task_count' in response_data:
                print(f'  Tasks Created/Updated: {response_data["task_count"]}')
            if 'message' in response_data:
                print(f'  Message: {response_data["message"]}')

            print('\nNext Steps:')
            print('1. Verify quest exists in Supabase (quests table)')
            print('2. Verify sample tasks created (quest_sample_tasks table)')
            print('3. Check that spark_assignment_id is set for each task')
            print('4. Test SSO login to verify auto-enrollment works')

            if 'quest_id' in response_data:
                print_verification_queries(response_data['quest_id'])

            return response_data

        else:
            print('\n✗ FAILED - Course sync webhook returned error')
            print(f'  Status: {response.status_code}')
            error_msg = response_data.get('error', 'Unknown error')
            print(f'  Error: {error_msg}')
            sys.exit(1)

    except requests.exceptions.RequestException as e:
        print(f'\n✗ ERROR - Failed to call course sync webhook')
        print(f'  Error: {str(e)}')
        sys.exit(1)


def print_verification_queries(quest_id):
    """Print SQL queries for manual verification"""
    print(f'\nVerification Queries (run in Supabase SQL Editor):')
    print('=' * 50)
    print(f'''
-- Verify quest created
SELECT
    id,
    title,
    description,
    quest_type,
    lms_course_id,
    lms_platform,
    is_active
FROM quests
WHERE id = '{quest_id}';

-- Verify sample tasks created
SELECT
    id,
    title,
    description,
    pillar,
    xp_value,
    spark_assignment_id,
    order_index
FROM quest_sample_tasks
WHERE quest_id = '{quest_id}'
ORDER BY order_index;

-- Check if quest has OnFire Learning material link
SELECT id, title, material_link
FROM quests
WHERE id = '{quest_id}';
''')


if __name__ == '__main__':
    test_course_sync_webhook()
    print('\nTest completed successfully!')

File: test_spark_course_sync.js
/**
 * Test SPARK Course Sync Webhook
 *
 * Tests the /spark/webhook/course endpoint to verify:
 * 1. Quest creation from course data
 * 2. Sample task creation from assignments
 * 3. HMAC signature validation
 */

const crypto = require('crypto');

// Configuration
const BACKEND_URL = process.env.BACKEND_URL || 'https://optio-dev-backend.onrender.com';
const SPARK_WEBHOOK_SECRET = process.env.SPARK_WEBHOOK_SECRET;

if (!SPARK_WEBHOOK_SECRET) {
    console.error('Error: SPARK_WEBHOOK_SECRET environment variable not set');
    process.exit(1);
}

// Test course data with assignments
const courseData = {
    spark_org_id: "24",  // Demo org ID from SPARK developer
    spark_course_id: "test_course_biology_101",
    course_title: "Biology 101 - Test Course",
    course_description: "Introduction to biology - automated test course with assignments",
    assignments: [
        {
            spark_assignment_id: "assign_bio_001",
            title: "Cell Structure Lab Report",
            description: "Complete the cell structure lab and submit a detailed report with diagrams",
            due_date: "2025-02-15T23:59:59Z"
        },
        {
            spark_assignment_id: "assign_bio_002",
            title: "Photosynthesis Essay",
            description: "Write a 500-word essay explaining the process of photosynthesis",
            due_date: "2025-02-22T23:59:59Z"
        },
        {
            spark_assignment_id: "assign_bio_003",
            title: "Ecosystem Project",
            description: "Create a presentation about a local ecosystem",
            due_date: "2025-03-01T23:59:59Z"
        }
    ]
};

// Calculate HMAC signature
function calculateSignature(payload) {
    const payloadString = JSON.stringify(payload, Object.keys(payload).sort());
    return crypto
        .createHmac('sha256', SPARK_WEBHOOK_SECRET)
        .update(payloadString)
        .digest('hex');
}

// Test course sync webhook
async function testCourseSyncWebhook() {
    console.log('Testing SPARK Course Sync Webhook');
    console.log('==================================\n');

    console.log('Backend URL:', BACKEND_URL);
    console.log('Course ID:', courseData.spark_course_id);
    console.log('Course Title:', courseData.course_title);
    console.log('Assignment Count:', courseData.assignments.length);
    console.log('\nAssignments:');
    courseData.assignments.forEach((assignment, idx) => {
        console.log(`  ${idx + 1}. ${assignment.title} (ID: ${assignment.spark_assignment_id})`);
    });

    // Calculate signature
    const signature = calculateSignature(courseData);
    console.log('\nCalculated HMAC signature:', signature.substring(0, 20) + '...');

    // Make request
    console.log('\nSending POST request to /spark/webhook/course...\n');

    try {
        const response = await fetch(`${BACKEND_URL}/spark/webhook/course`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-Spark-Signature': signature
            },
            body: JSON.stringify(courseData)
        });

        const responseText = await response.text();
        let responseData;

        try {
            responseData = JSON.parse(responseText);
        } catch (e) {
            console.error('Failed to parse response as JSON:', responseText);
            throw e;
        }

        console.log('Response Status:', response.status);
        console.log('Response Body:', JSON.stringify(responseData, null, 2));

        if (response.ok) {
            console.log('\n✓ SUCCESS - Course sync webhook processed successfully');
            console.log(`  Quest ID: ${responseData.quest_id}`);
            console.log(`  Tasks Created/Updated: ${responseData.task_count}`);
            console.log(`  Message: ${responseData.message}`);

            console.log('\nNext Steps:');
            console.log('1. Verify quest exists in Supabase (quests table)');
            console.log('2. Verify sample tasks created (quest_sample_tasks table)');
            console.log('3. Check that spark_assignment_id is set for each task');
            console.log('4. Test SSO login to verify auto-enrollment works');

            return responseData;
        } else {
            console.error('\n✗ FAILED - Course sync webhook returned error');
            console.error('  Status:', response.status);
            console.error('  Error:', responseData.error || 'Unknown error');
            throw new Error(`Course sync failed: ${responseData.error}`);
        }

    } catch (error) {
        console.error('\n✗ ERROR - Failed to call course sync webhook');
        console.error('  Error:', error.message);
        throw error;
    }
}

// Verify quest in database (optional - requires Supabase client)
async function verifyQuestCreated(questId) {
    console.log('\n\nVerifying Quest in Database');
    console.log('===========================\n');
    console.log('Quest ID:', questId);
    console.log('\nTo verify manually, run this query in Supabase:');
    console.log(`
SELECT
    q.id,
    q.title,
    q.description,
    q.quest_type,
    q.lms_course_id,
    q.is_active,
    COUNT(qst.id) as task_count
FROM quests q
LEFT JOIN quest_sample_tasks qst ON qst.quest_id = q.id
WHERE q.id = '${questId}'
GROUP BY q.id;

-- View tasks
SELECT
    id,
    title,
    description,
    pillar,
    xp_value,
    spark_assignment_id,
    order_index
FROM quest_sample_tasks
WHERE quest_id = '${questId}'
ORDER BY order_index;
    `);
}

// Run test
testCourseSyncWebhook()
    .then(result => {
        if (result && result.quest_id) {
            verifyQuestCreated(result.quest_id);
        }
        console.log('\n\nTest completed successfully!');
        process.exit(0);
    })
    .catch(error => {
        console.error('\n\nTest failed:', error.message);
        process.exit(1);
    });
